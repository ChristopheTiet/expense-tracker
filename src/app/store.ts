import { configureStore } from "@reduxjs/toolkit";
import expensesReducer from "../ducks/expenses-slice";
import filterReducer from "../ducks/filter-slice";

export const store = configureStore({
  reducer: {
    expenses: expensesReducer,
    filter: filterReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
