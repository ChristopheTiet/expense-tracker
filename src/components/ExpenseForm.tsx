import { Expense } from "../ducks/expenses-slice";
import moment from "moment";

interface ExpenseFormProps {
  expense?: Expense;
  onSubmit: (e: React.FormEvent) => void;
}

export default function ExpenseForm({ onSubmit, expense }: ExpenseFormProps) {
  return (
    <form
      onSubmit={(e) => {
        onSubmit(e);
      }}
    >
      <div className="flex space-x-2 items-baseline">
        <input
          type="text"
          name="description"
          required
          className="form-control block  px-3 py-1.5 text-base font-normal text-gray-700 bg-white bg-clip-padding border border-solid border-gray-300 rounded transition ease-in-out m-0 focus:text-gray-700 focus:bg-white focus:border-blue-600 focus:outline-none"
          placeholder="description"
          defaultValue={expense ? expense.description : ""}
        />
        <input
          type="number"
          name="amount"
          required
          step={0.01}
          className="form-control  block px-3 py-1.5 text-base font-normal text-gray-700 bg-white bg-clip-padding border border-solid border-gray-300 rounded transition ease-in-out m-0 focus:text-gray-700 focus:bg-white focus:border-blue-600 focus:outline-none"
          placeholder="amount"
          defaultValue={expense ? expense.amount : ""}
        />
        <textarea
          name="note"
          rows={1}
          className="form-control align-center items-center block px-3 py-1.5 text-base font-normal text-gray-700 bg-white bg-clip-padding border border-solid border-gray-300 rounded transition ease-in-out m-0 focus:text-gray-700 focus:bg-white focus:border-blue-600 focus:outline-none"
          placeholder="Note"
          defaultValue={expense ? expense.note : ""}
        />
        <input
          type="date"
          name="date"
          className="form-control align-center items-center block px-3 py-1.5 text-base font-normal text-gray-700 bg-white bg-clip-padding border border-solid border-gray-300 rounded transition ease-in-out m-0 focus:text-gray-700 focus:bg-white focus:border-blue-600 focus:outline-none"
          defaultValue={
            expense ? moment(expense.createdAt).format("YYYY-MM-DD") : ""
          }
        />
        <button
          type="submit"
          className="inline-block px-6 py-2.5 mb-6 bg-blue-600 text-white font-medium text-xs leading-tight uppercase rounded shadow-md hover:bg-blue-700 hover:shadow-lg focus:bg-blue-700 focus:shadow-lg focus:outline-none focus:ring-0 active:bg-blue-800 active:shadow-lg transition duration-150 ease-in-out"
        >
          {expense ? "Edit Expense" : "Add expense"}
        </button>
      </div>
    </form>
  );
}
