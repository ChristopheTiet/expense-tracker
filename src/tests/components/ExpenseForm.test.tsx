import renderer from "react-test-renderer";
import ExpenseForm from "../../components/ExpenseForm";
import { BrowserRouter } from "react-router-dom";
import { store } from "../../app/store";
import { Provider } from "react-redux";
import expenses from "../fixtures/expenses";
import { fireEvent, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

test("should render ExpenseForm correctly with no data", () => {
  const tree = renderer
    .create(
      <BrowserRouter>
        <Provider store={store}>
          <ExpenseForm
            onSubmit={() => {
              console.log("test");
            }}
          />
        </Provider>
      </BrowserRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("should render ExpenseForm correctly with data", () => {
  const tree = renderer
    .create(
      <BrowserRouter>
        <Provider store={store}>
          <ExpenseForm
            onSubmit={() => {
              console.log("test");
            }}
            expense={expenses[0]}
          />
        </Provider>
      </BrowserRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("should render that description input is invalid if empty form submission", () => {
  render(
    <ExpenseForm
      onSubmit={(e) => {
        e.preventDefault();
      }}
    />
  );
  userEvent.click(screen.getByText("Add expense"));
  expect(screen.getByPlaceholderText("description")).toBeInvalid;
});
