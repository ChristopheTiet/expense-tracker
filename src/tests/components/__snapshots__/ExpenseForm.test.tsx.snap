// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseForm correctly with data 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="flex space-x-2 items-baseline"
  >
    <input
      className="form-control block  px-3 py-1.5 text-base font-normal text-gray-700 bg-white bg-clip-padding border border-solid border-gray-300 rounded transition ease-in-out m-0 focus:text-gray-700 focus:bg-white focus:border-blue-600 focus:outline-none"
      defaultValue="cake"
      name="description"
      placeholder="description"
      required={true}
      type="text"
    />
    <input
      className="form-control  block px-3 py-1.5 text-base font-normal text-gray-700 bg-white bg-clip-padding border border-solid border-gray-300 rounded transition ease-in-out m-0 focus:text-gray-700 focus:bg-white focus:border-blue-600 focus:outline-none"
      defaultValue={15}
      name="amount"
      placeholder="amount"
      required={true}
      step={0.01}
      type="number"
    />
    <textarea
      className="form-control align-center items-center block px-3 py-1.5 text-base font-normal text-gray-700 bg-white bg-clip-padding border border-solid border-gray-300 rounded transition ease-in-out m-0 focus:text-gray-700 focus:bg-white focus:border-blue-600 focus:outline-none"
      defaultValue=""
      name="note"
      placeholder="Note"
      rows={1}
    />
    <input
      className="form-control align-center items-center block px-3 py-1.5 text-base font-normal text-gray-700 bg-white bg-clip-padding border border-solid border-gray-300 rounded transition ease-in-out m-0 focus:text-gray-700 focus:bg-white focus:border-blue-600 focus:outline-none"
      defaultValue="1970-01-01"
      name="date"
      type="date"
    />
    <button
      className="inline-block px-6 py-2.5 mb-6 bg-blue-600 text-white font-medium text-xs leading-tight uppercase rounded shadow-md hover:bg-blue-700 hover:shadow-lg focus:bg-blue-700 focus:shadow-lg focus:outline-none focus:ring-0 active:bg-blue-800 active:shadow-lg transition duration-150 ease-in-out"
      type="submit"
    >
      Edit Expense
    </button>
  </div>
</form>
`;

exports[`should render ExpenseForm correctly with no data 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="flex space-x-2 items-baseline"
  >
    <input
      className="form-control block  px-3 py-1.5 text-base font-normal text-gray-700 bg-white bg-clip-padding border border-solid border-gray-300 rounded transition ease-in-out m-0 focus:text-gray-700 focus:bg-white focus:border-blue-600 focus:outline-none"
      defaultValue=""
      name="description"
      placeholder="description"
      required={true}
      type="text"
    />
    <input
      className="form-control  block px-3 py-1.5 text-base font-normal text-gray-700 bg-white bg-clip-padding border border-solid border-gray-300 rounded transition ease-in-out m-0 focus:text-gray-700 focus:bg-white focus:border-blue-600 focus:outline-none"
      defaultValue=""
      name="amount"
      placeholder="amount"
      required={true}
      step={0.01}
      type="number"
    />
    <textarea
      className="form-control align-center items-center block px-3 py-1.5 text-base font-normal text-gray-700 bg-white bg-clip-padding border border-solid border-gray-300 rounded transition ease-in-out m-0 focus:text-gray-700 focus:bg-white focus:border-blue-600 focus:outline-none"
      defaultValue=""
      name="note"
      placeholder="Note"
      rows={1}
    />
    <input
      className="form-control align-center items-center block px-3 py-1.5 text-base font-normal text-gray-700 bg-white bg-clip-padding border border-solid border-gray-300 rounded transition ease-in-out m-0 focus:text-gray-700 focus:bg-white focus:border-blue-600 focus:outline-none"
      defaultValue=""
      name="date"
      type="date"
    />
    <button
      className="inline-block px-6 py-2.5 mb-6 bg-blue-600 text-white font-medium text-xs leading-tight uppercase rounded shadow-md hover:bg-blue-700 hover:shadow-lg focus:bg-blue-700 focus:shadow-lg focus:outline-none focus:ring-0 active:bg-blue-800 active:shadow-lg transition duration-150 ease-in-out"
      type="submit"
    >
      Add expense
    </button>
  </div>
</form>
`;
